File: upa-impl/src/main/csharp/Sources/Custom/Config/MakeSessionAwareMethodInterceptor2.cs
using Net.Vpc.Upa.Impl.Util;

namespace Net.Vpc.Upa.Impl.Config
{


    /**
    * @author Taha BEN Salah
    * @creationdate 1/7/13 2:32 PM*/
    internal class MakeSessionAwareMethodInterceptor2<T> {

        private readonly Net.Vpc.Upa.MethodFilter methodFilter;

        public MakeSessionAwareMethodInterceptor2(Net.Vpc.Upa.MethodFilter methodFilter) {
            this.methodFilter = methodFilter;
        }

        public virtual T MakeSessionAware() /* throws Net.Vpc.Upa.Exceptions.UPAException */
        {
            T t = (T)System.Activator.CreateInstance(typeof (T));
            return t.Proxy<T>(MyInterceptorMethod);
        }

        public object MyInterceptorMethod<T>(T@object, string methodName, object[] args, EmitProxyExecute<T> execute){
            if (methodFilter == null /*|| methodFilter.Accept(method)*/) {
                Net.Vpc.Upa.PersistenceGroup persistenceGroup = Net.Vpc.Upa.UPA.GetPersistenceGroup();
                Net.Vpc.Upa.Session s = null;
                try {
                    s = persistenceGroup.GetCurrentSession();
                } catch (Net.Vpc.Upa.Exceptions.CurrentSessionNotFoundException ignore) {
                }
                //ignore
                bool sessionCreated = false;
                if (s == null) {
                    s = persistenceGroup.OpenSession();
                    sessionCreated = true;
                }
                bool transactionCreated = false;
                persistenceGroup.GetPersistenceUnit().BeginTransaction(Net.Vpc.Upa.TransactionType.REQUIRED);
                transactionCreated = true;
                object ret = null;
                try {
                    ret = execute(@object, args);
                    persistenceGroup.GetPersistenceUnit().CommitTransaction();
                } catch (System.Exception e) {
                    if (transactionCreated) {
                        persistenceGroup.GetPersistenceUnit().RollbackTransaction();
                    }
                    throw e;
                } finally {
                    if (sessionCreated) {
                        s.Close();
                    }
                }
                return ret;
            } else {
                return execute(@object, args);
            }
        }
    }
}

File: upa-impl/src/main/java/net/vpc/upa/impl/config/MakeSessionAwareMethodInterceptor2.java
package net.vpc.upa.impl.config;

import net.vpc.upa.MethodFilter;
import net.vpc.upa.UPA;
import net.vpc.upa.impl.util.PlatformMethodProxy;
import net.vpc.upa.impl.util.PlatformMethodProxyEvent;

/**
 * @author Taha BEN SALAH <taha.bensalah@gmail.com>
 * @creationdate 1/7/13 2:32 PM
 */
class MakeSessionAwareMethodInterceptor2<T> implements PlatformMethodProxy<T> {

    private final MethodFilter methodFilter;

    public MakeSessionAwareMethodInterceptor2(MethodFilter methodFilter) {
        this.methodFilter = methodFilter;
    }

    @Override
    public Object intercept(PlatformMethodProxyEvent<T> event) throws Throwable {
        if (methodFilter == null || methodFilter.accept(event.getMethod())) {
            return UPA.getContext().invoke(new MakeSessionAwareMethodInterceptor2Action(event), null);
        } else {
            return event.invokeBase(event.getObject(), event.getArguments());
        }
    }

}
